### PR TITLE
Issue #134: Add automatic image derivatives.

### PR DESCRIPTION
**Pull request by e0ipso**
Implements #134 as an alternative to #142.

@amitaibu this also has the nice side effect to allow passing arguments to `'callback'` and `'process_callbacks'`.

### CODE REVIEW

**Comment by e0ipso** [comment]
@amitaibu I think this is the last feature that we need to support before an stable release. Can you think of any other must have?

**Comment by e0ipso** [comment]
W00t! Tests are green in the first try.

File: modules/restful_example/plugins/restful/node/articles/1.5/articles__1_5.inc [hunk 1]
```diff
@@ -11,4 +11,9 @@ $plugin = array(
   'authentication_types' => TRUE,
   'authentication_optional' => TRUE,
   'minor_version' => 5,
+  'options' => array(
+    'image_styles' => array(
```
Comment (amitaibu) [inline modules/restful_example/plugins/restful/node/articles/1.5/articles__1_5.inc]: shouldn't we move this to the fieldsInfo() array, OR, allow overriding these plugin level ones, with field level ones?

File: plugins/restful/RestfulBase.php [hunk 2]
```diff
@@ -315,6 +315,37 @@ public function format(array $data) {
...
+   *     - An array containing an object and a method name of that object.
+   *       Ex: array($this, 'format').
+   *     - An array containing any of the methods before and an array of
+   *       parameters to pass to the callback.
+   *       Ex: array(array($this, 'processing'), array('param1', 2))
+   * @param array $params
+   *   Array of additional parameters to pass in.
+   *
+   * @return mixed
+   *   Whatever the callback returns.
```
Comment (amitaibu) [inline plugins/restful/RestfulBase.php]: The return value of the callback.

File: plugins/restful/RestfulBase.php [hunk 3]
```diff
@@ -315,6 +315,37 @@ public function format(array $data) {
...
+   *
+   * @throws \RestfulException
+   */
+  public static function executeCallback($callback, array $params = array()) {
+    if (!is_callable($callback)) {
+      if (is_array($callback) && count($callback) == 2 && is_array($callback[1])) {
+        return static::executeCallback($callback[0], array_merge($params, $callback[1]));
```
Comment (amitaibu) [inline plugins/restful/RestfulBase.php:343]: I understand it, but worth a doc with example.

File: plugins/restful/RestfulBase.php [hunk 4]
```diff
@@ -315,6 +315,37 @@ public function format(array $data) {
...
+      }
+      $callback_name = is_array($callback) ? $callback[1] : $callback;
+      throw new \RestfulException(format_string('Process callback function: @callback does not exists.', array('@callback' => $callback_name)));
```
Comment (amitaibu) [inline plugins/restful/RestfulBase.php]: `Process callback` should probably be changed to just `callback`

File: plugins/restful/RestfulEntityBase.php [hunk 5]
```diff
@@ -1337,9 +1324,15 @@ public function getPublicFields() {
         'column' => FALSE,
       );
 
+      // Get the image styles.
+      $options = $this->getPluginInfo('options');
       if ($info['property'] && !$info['column'] && $field = field_info_field($info['property'])) {
         // Set the column name.
         $info['column'] = key($field['columns']);
+        // If it's an image check if we need to add image style processing.
```
Comment (amitaibu) [inline plugins/restful/RestfulEntityBase.php:1327]: this should be outside of the IF statement (as it won't get in if `$info['column']` was explicitly declared).

File: plugins/restful/RestfulEntityBase.php [hunk 6]
```diff
@@ -1373,4 +1366,20 @@ public function isListRequest() {
     return empty($path) || strpos($path, ',') !== FALSE;
   }
 
+  /**
+   * Get the image variants based on the configured styles.
```
Comment (amitaibu) [inline plugins/restful/RestfulEntityBase.php]: Get the image URLs based on the configured image styles.

**Comment by e0ipso** [comment]
@amitaibu this is ready.

File: README.md [hunk 7]
```diff
@@ -56,6 +56,50 @@ $handler->patch($id, $request);
...
+By default the RESTful module will expose the ID, label and URL of the entity.
+You probably want to expose more than that. To do so you will need to implement
+the `publicFieldsInfo` method defining the names in the output array and how
+those are mapped to the queried entity. For instance the following example will
+retrieve the basic fields plus the body, tags and images from (probably) an
+article node. The RESTful module will know to use the `MyRestfulPlugin` class
+because your plugin definition will say so.
+
+```php
+class MyRestfulPlugin extends \RestfulEntityBase {
```
Comment (amitaibu) [inline README.md]: Maybe `MyRestfulPlugin` => `MyArticlesResource` ("best practice" naming convention)

File: plugins/restful/RestfulEntityBase.php [hunk 8]
```diff
@@ -1373,4 +1365,27 @@ public function isListRequest() {
     return empty($path) || strpos($path, ',') !== FALSE;
   }
 
+  /**
+   * Get the image URLs based on the configured image styles.
+   *
+   * @param array $file_array
+   *   The file array from EMW.
```
Comment (amitaibu) [inline plugins/restful/RestfulEntityBase.php]: => `The file array.`

File: README.md [hunk 9]
```diff
@@ -56,6 +56,50 @@ $handler->patch($id, $request);
 ```
 
 ### View an entity
+By default the RESTful module will expose the ID, label and URL of the entity.
+You probably want to expose more than that. To do so you will need to implement
+the `publicFieldsInfo` method defining the names in the output array and how
+those are mapped to the queried entity. For instance the following example will
+retrieve the basic fields plus the body, tags and images from (probably) an
```
Comment (amitaibu) [inline README.md]: we can probably remove the word "probably" :)

**Comment by amitaibu** [comment]
Minor comments, after that can be merged.

File: README.md (hunk 9 above)
Comment (e0ipso) [inline README.md]: LOL, you're probably right.

**Comment by e0ipso** [comment]
Merge squashed in 93b0b4db4c2b0f803715468ac464cbcd2e57fd34.

**Pull request closed without merging**